### PR TITLE
Adding Word Iteration Functionality

### PR DESCRIPTION
**Pull request by tunystom**
This adds a possibility to iterate over words within the same line. The old functionality remained intact.

### CODE REVIEW

File: tests/dump_file.py [hunk 1]
```diff
@@ -32,12 +34,21 @@
...
+            print '  ' * 1, 'Flow'
             for block in flow:
-                print ' ' * 2, 'Block', '(bbox: %0.2f, %0.2f, %0.2f, %0.2f)' % block.bbox.as_tuple()
+                print '  ' * 2, 'Block', '(bbox: %0.2f, %0.2f, %0.2f, %0.2f)' % block.bbox.as_tuple()
                 for line in block:
-                    print ' ' * 3, 'Line', line.text.encode('UTF-8'), '(bbox: %0.2f, %0.2f, %0.2f, %0.2f)' % line.bbox.as_tuple()
-                    if args.chars:
+                    print '  ' * 3, 'Line',
+                    if not args.words:
+                        line.text.encode('UTF-8'),
```
Comment (bzamecnik) [inline tests/dump_file.py]: Maybe `print line.text.encode('UTF-8'),`?

File: pdfparser/poppler.pyx [hunk 2]
```diff
@@ -575,134 +577,211 @@ cdef class LinesIterator:
...
+                continue
+
+            bboxes = word.char_bboxes
+
+            # if the previous word ended with space,
+            # it needs to be added because it is not
+            # a part of the word
+            if add_word_space:
+                word_texts.append(u' ')
+                char_bboxes.append(BBox(char_bboxes[-1].x2,
```
Comment (bzamecnik) [inline pdfparser/poppler.pyx]: What about `char_bboxes[-1]` in the first iteration when `char_bboxes` is empty?

Comment (bzamecnik) [reply pdfparser/poppler.pyx]: Aha, that can happen only if `add_word_space == True`, ie. not in the first iteration.

Comment (bzamecnik) [reply pdfparser/poppler.pyx]: Would be perfect to comment that space left = previous word right, space right = current word left. Possibly extract `prev_word_bbox = char_bboxes[-1]` and `word_bbox = bboxes[0]`.

File: pdfparser/poppler.pyx [hunk 3]
```diff
@@ -575,134 +577,211 @@ cdef class LinesIterator:
...
+    def __cinit__(self, PopplerPage page):
+        self.page = page
+        self.word = NULL
+
+    cdef Word wrap(self, TextWord *word):
+        cdef:
+            GooString *_text
+            GooString *_font_name
+            int        i, word_length
+            double     l, t, r, b, g
```
Comment (bzamecnik) [inline pdfparser/poppler.pyx]: Possibly change to `left, top, right, bottom, red, green, blue` or at least comment that `r` is reused for `right`/`red` and `b` for `bottom`/`blue` 🎃

File: pdfparser/poppler.pyx [hunk 4]
```diff
@@ -575,134 +577,211 @@ cdef class LinesIterator:
...
+        # sanity check for data integrity
+        assert len(self.text) == len(self.char_bboxes)
+        assert len(self.text) == len(self.char_fonts)
+
+        return self
 
     property bbox:
-        '''
-        The bounding box of this line of text
-        as (left, top, right, bottom).
```
Comment (bzamecnik) [inline pdfparser/poppler.pyx:690]: Would be good to move to the `classs BBox` so that we know the x/y convention.

**Review by bzamecnik — CHANGES_REQUESTED** [review]
Good job. Missing print and a few code style enhancements.

File: pdfparser/poppler.pyx (hunk 4 above)
Comment (tunystom) [reply pdfparser/poppler.pyx:690]: I added this comment there as you suggested.

**Review by bzamecnik — APPROVED** [review]

**Comment by bzamecnik** [comment]
LGTM. Thanks.

**Comment by bzamecnik** [comment]
I tried to build it and run it on some invoice and it correctly shows words.

```
 python tests/dump_file.py foo.pdf --words
```

**Pull request merged**